design/tlog-spilling.md.html

```diff
@@ -0,0 +1,590 @@
+
+Popping works similarly to before, but now requires recovering information from
+disk.  Previously, we would maintain a map from version to location in the disk
+queue for every version we hadn't yet spilled.  Once spilling has copied the
+value into the B-tree, knowing where the commit was in the disk queue is
+useless to us, and is removed.  In spill-by-reference, that information is
+still needed to know how to map "pop until version 7" to "pop until byte 87" in
+the disk queue.  Unfortunately, keeping this information in memory would result
+in TLogs slowly consuming more and more memory[^versionmap-memory] as more data
+is spilled.  Instead, we issue a range read of the B-tree from `(tag, pop_version)`
+to `(tag, infinity)` and look at the first commit we find with a version
+greater than our own.  We then use its starting disk queue location as the
+limit of what we could pop the disk queue until for this tag.
+
+[^versionmap-memory]: Pessimistic assumptions would suggest that a TLog spilling 1TB of data would require ~50GB of memory to hold this map, which isn't acceptable.
+
+## Detailed Implementation
+
+The rough outline of concrete changes proposed looks like:
+
+0. Allow a new TLog and old TLog to co-exist and be configurable, upgradeable, and recoverable
+0. Modify spilling in new TLogServer
+0. Modify peeking in new TLogServer
+0. Modify popping in new TLogServer
+0. Spill txsTag specially
+
+### Spilling
+
+In spill-by-reference, spilling is now the act of persisting the index of 
+
+### Peeking
+
+A `TLogPeekRequest` contains a `Tag` and a `Version`, and is a request for all
+commits with the specified tag with a commit version greater than or equal to
+the given version.  The goal is to return a 150KB block of mutations.
+
+**************************************************************************
+*                                                                        *
+* +---------+  Tag    +---------+           Tag    +--------+            *
+* |  Peek   |-------->| Spilled | ...------------->| Memory |            *
+* | Request | Version |  Index  |          Version | Index  |            *
+* +---------+         +---------+                  +--------+            *
+*                        |                                 |             *
+*      +-----------------+-----------------+               |             *
+*     / \  Start=100   _/ \_  Start=500    +  Start=900    +  Ptr=0xF00  *
+*    /   \ Length=50  /     \ Length=70   / \ Length=30   / \ Length=30  *
+*  +------------------------------------------------+------------------+ *
+*  |                   Disk Queue                   |  Also In Memory  | *
+*  +------------------------------------------------+------------------+ *
+*                                                                        *
+**************************************************************************
+
+Spill-by-value and memory storage engine only ever read from DiskQueue when
```

Comment:
This makes a lot of sense but I would probably bring it up a lot sooner as it really helps explain the difference between the in-memory queues and the disk queue and when each is used. Also you say "when recovering" and I assume that is specifically referring to when TLog node is recovering, not when a storage node is recovering